subsys/bluetooth/host/gatt.c

```diff
@@ -2110,47 +2078,74 @@ static bool gatt_cf_notify_multi(struct bt_conn *conn)
 }
 
-static int gatt_notify_mult(struct bt_conn *conn, uint16_t handle,
+static int gatt_notify_mult(struct bt_conn *conn, uint16_t num_params,
 			    struct bt_gatt_notify_params *params)
 {
-	struct net_buf **buf = &nfy_mult[bt_conn_index(conn)];
+	struct net_buf *buf;
 	struct bt_att_notify_mult *nfy;
+	struct notify_data data;
+	uint16_t pdu_len = 0;
+
+	for (uint16_t i = 0; i < num_params; i++) {
+		data.attr = params[i].attr;
+		data.handle = bt_gatt_attr_get_handle(params[i].attr);
+		/* Lookup UUID if it was given */
+		if (params[i].uuid) {
+			if (!gatt_find_by_uuid(&data, params[i].uuid)) {
+				return -ENOENT;
+			}
+			params[i].attr = data.attr;
+		} else {
+			if (!data.handle) {
+				return -ENOENT;
+			}
+		}
 
-	/* Check if we can fit more data into it, in case it doesn't fit send
-	 * the existing buffer and proceed to create a new one
-	 */
-	if (*buf && ((net_buf_tailroom(*buf) < sizeof(*nfy) + params->len) ||
-	    !bt_att_tx_meta_data_match(*buf, params->func, params->user_data))) {
-		int ret;
-
-		ret = gatt_notify_mult_send(conn, buf);
-		if (ret < 0) {
-			return ret;
+		/* Confirm that the connection has the correct level of security */
+		if (bt_gatt_check_perm(conn, params[i].attr,
+				       BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_READ_AUTHEN)) {
+			BT_WARN("Link is not encrypted");
+			return -EPERM;
 		}
-	}
 
-	if (!*buf) {
-		*buf = bt_att_create_pdu(conn, BT_ATT_OP_NOTIFY_MULT,
-					 sizeof(*nfy) + params->len);
-		if (!*buf) {
-			return -ENOMEM;
+		/* Calculate the ATT_MULTIPLE_HANDLE_VALUE_NTF pdu length */
+		pdu_len += sizeof(*nfy) + params[i].len;
+		if (pdu_len > UINT16_MAX) {
+			return -EINVAL;
 		}
+	}
 
-		bt_att_set_tx_meta_data(*buf, params->func, params->user_data);
+	buf = bt_att_create_pdu(conn, BT_ATT_OP_NOTIFY_MULT, pdu_len);
+	if (!buf) {
+		return -ENOMEM;
 	}
 
-	BT_DBG("handle 0x%04x len %u", handle, params->len);
+	for (uint16_t i = 0; i < num_params; i++) {
+		data.handle = bt_gatt_attr_get_handle(params[i].attr);
 
-	nfy = net_buf_add(*buf, sizeof(*nfy));
-	nfy->handle = sys_cpu_to_le16(handle);
-	nfy->len = sys_cpu_to_le16(params->len);
+		/* Check if attribute is a characteristic then adjust the handle */
+		if (!bt_uuid_cmp(params[i].attr->uuid, BT_UUID_GATT_CHRC)) {
+			struct bt_gatt_chrc *chrc = params[i].attr->user_data;
 
-	net_buf_add(*buf, params->len);
-	memcpy(nfy->value, params->data, params->len);
+			if (!(chrc->properties & BT_GATT_CHRC_NOTIFY)) {
+				return -EINVAL;
+			}
 
-	k_work_submit(&nfy_mult_work);
+			data.handle = bt_gatt_attr_value_handle(params[i].attr);
+		}
 
-	return 0;
+		BT_DBG("handle 0x%04x len %u", data.handle, params[i].len);
+
+		nfy = net_buf_add(buf, sizeof(*nfy));
+		nfy->handle = sys_cpu_to_le16(data.handle);
+		nfy->len = sys_cpu_to_le16(params[i].len);
+
+		net_buf_add(buf, params[i].len);
+		(void)memcpy(nfy->value, params[i].data, params[i].len);
+	}
+
+	bt_att_set_tx_meta_data(buf, params[0].func, params[0].user_data);
```

Comment:
Indeed. All the individual `params` should have the callback invoked. But where do we store the whole list of `params`? We could..
 - ..add a `_internal.callback_chain` field to `struct bt_gatt_notify_params`. But then we again use the `params` struct to store temporary data, which as commented further up, is not great design.
 - ..require that the `bt_gatt_notify_params` are wrapped in a new struct specialized for multi-notifications.
 - ..somehow manage the storage for it in gatt.c.

Reply:
It is OK for me, as long as either create a TODO or a GH issue that documents the wish to improve this. 

My biggest worry with this API (which pretty much also existing beforehand), is that as an application, actually using the `bt_gatt_notify_multiple` instead of just calling `bt_gatt_notify` multiple time is so much more of a hassle, as well as being more limited that I do not have many intentions of actually using it as is (but at least we'll fix the issues the current implementation has).

Reply:
yep agreed, the multiple api is a bit useless right now, I think we'll wait for actual use cases before reworking it, does that sound reasonable ? I'd like to get some feedback on how the users intend on using it, then we can make the necessary adjustments.